source/dem/localize_contacts.cc

```diff
@@ -27,216 +28,196 @@ localize_contacts(
-       particle_floating_wall_pairs_iterator !=
-       particle_floating_wall_in_contact.end();
-       ++particle_floating_wall_pairs_iterator)
-    {
-      auto particle_id = particle_floating_wall_pairs_iterator->first;
-
-      auto pairs_in_contant_content =
-        &particle_floating_wall_pairs_iterator->second;
+              // Get the reference to all the 2nd particle contact
+              // candidates of the current particle from the new board search
+              auto object_contact_candidates = &contact_candidates[object_id];
 
-      for (auto particle_floating_wall_map_iterator =
-             pairs_in_contant_content->begin();
-           particle_floating_wall_map_iterator !=
-           pairs_in_contant_content->end();)
-        {
-          auto floating_wall_id = particle_floating_wall_map_iterator->first;
-          auto particle_floating_wall_contact_candidate_element =
-            &particle_floating_wall_contact_candidates[particle_id];
+              // Check if the adjacent particle from history is a
+              // particle candidate of the current particle and get the
+              // iterator to that object if so
+              auto search_iterator =
+                std::find(contact_candidate_element->begin(),
+                          contact_candidate_element->end(),
+                          object_id);
 
-          auto search_iterator =
-            particle_floating_wall_contact_candidate_element->find(
-              floating_wall_id);
+              // Check if the current particle is a particle candidate
+              // of the adjacent particle in history
+              auto search_object_to_particle_iterator =
+                std::find(object_contact_candidates->begin(),
+                          object_contact_candidates->end(),
+                          particle_id);
 
-          if (search_iterator !=
-              particle_floating_wall_contact_candidate_element->end())
-            {
-              particle_floating_wall_contact_candidate_element->erase(
-                search_iterator);
-              ++particle_floating_wall_map_iterator;
+              // Particle-particle pairs are removed from the list of the new
+              // board search candidates or from the list of the history of the
+              // fine search candidate list depending on the last checks
+              if (search_iterator != contact_candidate_element->end())
+                {
+                  // The current 2nd is a candidate to the current particle.
+                  // The 2nd particle is then removed from the new board search
+                  // particle-particle contact candidates list but kept as
+                  // history in the fine search list
+                  contact_candidate_element->erase(search_iterator);
+                  ++adjacent_map_iterator;
+                }
+              else if (search_object_to_particle_iterator !=
```

Comment:
Make these if const expr by templating the Contact Type. You want to avoid ifs at run time in these critical structures